tuning.py

```
@@ -0,0 +1,317 @@
+    """Returns average effect size of singles for a given threshold.
+
+    Args:
+      landscape: The landscape.
+      threshold: The threshold fitness for singles.
+      adaptive: If True, thresholded singles will have fitness >= `threshold`. If False,
+        selected singles will have fitness < `threshold`.
+
+    Returns:
+      The average effect size of the selected singles.
+    """
+    all_singles = sampling.get_all_single_mutants(landscape.wildtype_sequence, landscape.vocab_size)
+    df = experiment.get_fitness_df(all_singles, landscape.evaluate, landscape.wildtype_sequence)
+    if adaptive:
+        mean_single_effect = df[df.fitness >= threshold].fitness.mean()
+    else:
+        mean_single_effect = df[df.fitness < threshold].fitness.mean()
+    return mean_single_effect
+
+
+def get_epistatic_horizon(landscape: potts_model.PottsModel) -> float:
+    r"""Returns the epistatic horizon for the given landscape.
+
+    The "epistatic horizon" is defined as the distance K from the wildtype at which, on average,
+    epistatic contributions outweigh linear contributions from adaptive singles. This is the average
+    distance we can expect a greedy algorithm to perform well on the given landscape.
+
+    Let $s_+$ be the average adaptive single mutant effect. let $e_{+,+}$ be the average
+    epistatic effect for a pair of adaptive singles. Then K is defined as :
+
+    $$
+    K * s_+ + (K \\choose 2) e_{+,+} = 0
+    $$
+
+    Solving for K
+    $$
+    K = \\dfrac{e_{+,+} - 2 s_+}
+               {e_{+,+}}
+    $$
+
+    Returns:
+      The epistatic horizon.
+    """
+    mean_adaptive_epistasis, _ = get_epistasis_stats(landscape, threshold=0.0, adaptive=True)
+    mean_adaptive_single = get_mean_single_effect(landscape, threshold=0, adaptive=True)
+    epistatic_horizon = (mean_adaptive_epistasis - 2 * mean_adaptive_single) / mean_adaptive_epistasis
```

what if the denominator is 0? return np.inf

---

I'd consider raising an error in this case. Likely there is something weird going on upstream.

---

A linear landscape will have a denominator of 0, which doesn't seem like it should raise an error...